src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -75,6 +75,12 @@ internal class DiagnosticComputer
                 return SerializableDiagnosticAnalysisResults.Empty;
             }
 
+            if (_document == null && analyzers.Length < compilationWithAnalyzers.Analyzers.Length)
+            {
+                // PERF: Generate a new CompilationWithAnalyzers with trimmed analyzers for non-document analysis case.
+                compilationWithAnalyzers = compilationWithAnalyzers.Compilation.WithAnalyzers(analyzers, compilationWithAnalyzers.AnalysisOptions);
+            }
```

Comment:
couldn't we always just do this, and then have .WithAnalyzers smartly no-op if passing in the same analyzers it poitns at?

Reply:
That would need a compiler change and possibly a new compiler public API `WithAnalyzer(analyzers)` that does not take options. Also note that for open file analysis we explicitly want to avoid this fork as we want to share the same `CompilationWithAnalyzers` instance across diagnostic computation calls.